egs/wsj/s5/steps/nnet3/libs/train_lib.py

```diff
@@ -0,0 +1,491 @@
+
+
+# Copyright 2016 Vijayaditya Peddinti.
+#           2016 Vimal Manohar
+# Apache 2.0.
+
+# This is a module with methods which will be used by scripts for training of
+# deep neural network acoustic model and raw model (i.e., generic neural
+# network without transition model) with frame-level objectives.
+
+import logging
+import math
+import imp
+import os
+
+common_train_lib = imp.load_source('ntl', 'steps/nnet3/libs/common_train_lib.py')
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+handler = logging.StreamHandler()
+handler.setLevel(logging.INFO)
+formatter = logging.Formatter('%(asctime)s [%(filename)s:%(lineno)s - %(funcName)s - %(levelname)s ] %(message)s')
+handler.setFormatter(formatter)
+logger.addHandler(handler)
+
+def TrainNewModels(dir, iter, srand, num_jobs,
+                   num_archives_processed, num_archives,
+                   raw_model_string, egs_dir,
+                   left_context, right_context,
+                   momentum, max_param_change,
+                   shuffle_buffer_size, minibatch_size,
+                   cache_read_opt, run_opts,
+                   frames_per_eg = -1, min_deriv_time = None):
+    # We cannot easily use a single parallel SGE job to do the main training,
+    # because the computation of which archive and which --frame option
+    # to use for each job is a little complex, so we spawn each one separately.
+    # this is no longer true for RNNs as we use do not use the --frame option
+    # but we use the same script for consistency with FF-DNN code
+
+    chunk_level_training = False if frames_per_eg > 0 else True
+    deriv_time_opts = (""
+                       if min_deriv_time is None
+                       else "--optimization.min-deriv-time={0}".format(min_deriv_time)
```

Comment:
Actually without self-repair in ClipGradientComponent, it was always the backward direction that diverges after a few iterations (and I think it can be explained by the fact that we did not impose max-deriv-time so the backprop would take long time steps); with self-repair in ClipGradientComponent, the backward's divergence can be fixed. But then the forward divergence gradually happens.